module/executiondatasync/optimistic_syncing/core.go

```diff
@@ -16,4 +29,238 @@ type Core interface {
+		executionResult,
+		header,
+	)
+
+	requesterComponent, err := requester.NewOneshotExecutionDataRequester(
+		coreLogger,
+		execDataRequesterMetrics,
+		downloader,
+		executionResult,
+		header,
+		execDataRequesterConfig,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &CoreImpl{
+		log:                      coreLogger,
+		execDataRequester:        requesterComponent,
+		txResultErrMsgsRequester: tx_error_messages.NewTransactionErrorMessagesRequester(txErrCore, &txErrMsgRequesterConfig, executionResult),
+		indexer:                  indexerComponent,
+		persister:                persisterComponent,
+		executionResult:          executionResult,
+		header:                   header,
+		registers:                registers,
+		events:                   events,
+		collections:              collections,
+		transactions:             transactions,
+		results:                  results,
+		txResultErrMsgs:          txResultErrMsgs,
+		protocolDB:               protocolDB,
+	}, nil
+}
+
+// Download implements the Core.Download method.
+// It downloads execution data for the block and stores it in the execution data cache.
+func (c *CoreImpl) Download(ctx context.Context) error {
+	blockID := c.executionResult.BlockID
+	height := c.header.Height
+
+	c.log.Debug().
+		Hex("block_id", logging.ID(blockID)).
+		Uint64("height", height).
+		Msg("downloading execution data")
+
+	executionData, err := c.execDataRequester.RequestExecutionData(ctx)
+	if err != nil {
+		return err
+	}
+
+	c.executionData = execution_data.NewBlockExecutionDataEntity(c.executionResult.ExecutionDataID, executionData)
+
+	//TODO: this should return []flow.TransactionResultErrorMessage and initialize txResultErrMsgsData
+	err = c.txResultErrMsgsRequester.RequestTransactionErrorMessages(ctx)
+	if err != nil {
+		return err
+	}
+
+	//TODO: initialize txResultErrMsgsData with returned data from requester
+	c.txResultErrMsgsData = make([]flow.TransactionResultErrorMessage, 0)
+
+	return nil
+}
+
+// Index implements the Core.Index method.
+// It retrieves the downloaded execution data from the cache and indexes it into in-memory storage.
+func (c *CoreImpl) Index(ctx context.Context) error {
+	blockID := c.executionResult.BlockID
+
+	c.log.Debug().
+		Hex("block_id", logging.ID(blockID)).
+		Uint64("height", c.header.Height).
+		Msg("indexing execution data")
+
+	if err := c.indexer.IndexBlockData(c.executionData); err != nil {
+		return err
+	}
+
+	if err := c.indexer.IndexTxResultErrorMessagesData(c.txResultErrMsgsData); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Persist implements the Core.Persist method.
+// It persists the indexed data to permanent storage atomically.
+func (c *CoreImpl) Persist(ctx context.Context) error {
+	c.log.Debug().
+		Hex("block_id", logging.ID(c.executionResult.BlockID)).
+		Uint64("height", c.header.Height).
+		Msg("persisting execution data")
+
+	// Create a batch for atomic updates
+	batch := c.protocolDB.NewBatch()
```

Comment:
I also mentioned it in the persister PR, but I think the batch handling should go in the persister